### PR TITLE
Tune int8 AMX WoQ micro-kernel for CPU

### LINKED ISSUE #134449: [Inductor] torchchat segmentation fault in int8 using max-autotune
### 🐛 Describe the bug

Torchchat of int8 woq will encounter segmentation fault in https://github.com/pytorch/pytorch/commit/f951fcd1d7c4e991d1c9ef642fe7761d7104cda2 when using `max-autotune` in torch.compile.

Reproducer:

Basic README is in https://github.com/pytorch/torchchat?tab=readme-ov-file#installation .
git clone https://github.com/pytorch/torchchat.git
cd torchchat
install dependencies
./install_requirements.sh # This will download CUDA version of Torch, you can wait, and finally it will install 20240814 cpu nightly
Uninstall torch (default is 20240814 nightly, the above commit is not included). Install 20240815 nightly or build from source.

TORCHINDUCTOR_FREEZING=1 python3 torchchat.py generate llama3.1 --prompt 'Hello my name is' --quantize '{"linear:int8": {"bitwidth": 8, "groupsize": 0}}' --compile  --num-samples 5 --device cpu

Need the following temp diff in PyTorch:
```
$ git diff
diff --git a/torch/_inductor/decomposition.py b/torch/_inductor/decomposition.py
index d02d41c8212..436c57451d8 100644
— a/torch/_inductor/decomposition.py
+++ b/torch/_inductor/decomposition.py
@@ -282,7 +282,7 @@ def mm(
 ) -> torch.Tensor:
     # Our matrix vector multiplies only achieve peak bandwidth with coordinate descent tuning.
     # todo: Look into why and fix it (hopefully)

   if config.coordinate_descent_tuning:
+    if config.coordinate_descent_tuning and self.device.type != "cpu":
         if guard_size_oblivious(self.shape[0] == 1) or guard_size_oblivious(
             input2.shape[1] == 1
         ):
```

Need to use `max-autotune` in torchchat:
https://github.com/pytorch/torchchat/blob/main/generate.py#L670

```
self.decode_one_token = torch.compile(
      self.decode_one_token, mode="max-autotune", fullgraph=True
)
```

### Versions

1

### PR DESCRIPTION
**Pull request by sanchitintel**
This patch prevents performance regression against the default ATen implementation for LLaMA 3.1 int8 GPTQ WoQ workload.

Uses AMX micro-kernel only if `M` >= `block_m`

cc @frank-wei @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by sanchitintel** [comment]
@jgong5, maybe it'd be better to disable the int8 AMX micro-kernel in case of a low `M`, instead of disabling it altogether.
Even on CentOS, AMX performs better if `M` >= `block_m`.

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 1]
```diff
@@ -603,15 +603,12 @@ class CppMicroGemmAMX(CppMicroGemm):
 
 {%- if input_dtype == torch.bfloat16 and input2_dtype == torch.int8 %}
     // create a buffer for tiles of B.
-    // TODO: loop-unrolling of the "compute" lambda may result in incorrect output
-    // as this buffer would be used, so maybe 4 of these should be used?
-    // Since UT output is correct, looks like loop unrolling isn't actually happening.
-    alignas(64) {{input_t}} bf16_weights_buf[512];
+    // alignment of 128 is being used to prevent segfault on RHEL with some specific libstdc++ version.
+    alignas(128) {{input_t}} bf16_weights_buf[512];
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_micro_gemm.py:607]: I suspect this is our local test environment issue. Better not "work around" it before understanding the root cause of it.

File: torch/_inductor/codegen/cpp_micro_gemm.py [hunk 2]
```diff
@@ -801,6 +798,14 @@ def create_from_config(cls, config: CppMicroGemmConfig):
                 ):
                     continue
                 block_m, block_n, block_k = config.register_blocking
+                if (
+                    config.vec_isa_cls == VecAMX
+                    and m < block_m
+                    and input_dtype == torch.bfloat16
+                    and input2_dtype == torch.int8
+                ):
+                    # For int8 WoQ GEMM, AMX micro-kernel may not perform well if m < block_m
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp_micro_gemm.py:807]: In that case, is AVX512 micro-gemm performs better than ATen?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/codegen/cpp_micro_gemm.py (hunk 2 above)
Comment (sanchitintel) [reply torch/_inductor/codegen/cpp_micro_gemm.py:807]: Yes

**Review by jgong5 — APPROVED** [review]

**Comment by sanchitintel** [comment]
Sorry, @jgong5! This branch is of PyTorch repo and not my fork, so there'd be a problem merging this one, so I created #135190 instead.

**Pull request closed without merging**